json-rpc/src/stream_rpc/subscriptions.rs

```diff
@@ -0,0 +1,258 @@
+    pub poll_interval_ms: u64,
+    pub queue_size: usize,
+}
+
+type JitterBackoff = Map<ExponentialBackoff, fn(Duration) -> Duration>;
+
+pub fn create_backoff(poll_interval_ms: u64) -> JitterBackoff {
+    ExponentialBackoff::from_millis(poll_interval_ms)
+        .factor(1000)
+        .max_delay(tokio::time::Duration::from_millis(poll_interval_ms * 10))
+        .map(jitter)
+}
+
+#[derive(Clone)]
+pub struct SubscriptionHelper {
+    pub db: Arc<dyn DbReader>,
+    pub client: ClientConnection,
+    pub jsonrpc_id: Id,
+    pub method: Method,
+    backoff: JitterBackoff,
+}
+
+impl SubscriptionHelper {
+    pub fn new(
+        db: Arc<dyn DbReader>,
+        client: ClientConnection,
+        jsonrpc_id: Id,
+        method: Method,
+    ) -> Self {
+        let poll_interval_ms = client.config.poll_interval_ms;
+        Self {
+            db,
+            client,
+            jsonrpc_id,
+            method,
+            backoff: create_backoff(poll_interval_ms),
+        }
+    }
+    /// This is to be one of the first things that a new subscription task must do, to let the client
+    /// know that the subscription has been created (some streams may be very sparse, or start with 0 items)
+    pub async fn send_ok(&self) {
+        counters::SUBSCRIPTION_OKS
+            .with_label_values(&[
+                self.client.connection_context.transport.as_str(),
+                &self.method.as_str(),
+                counters::RpcResult::Success.as_str(),
+                self.client.connection_context.sdk_info.language.as_str(),
+                &self.client.connection_context.sdk_info.version.to_string(),
+            ])
+            .inc();
+        self.client
+            .send_success(
+                self.jsonrpc_id.clone(),
+                &SubscribeResult::ok(self.db.get_latest_version().unwrap_or(0)),
+            )
+            .await
+            .ok();
+    }
+
+    pub fn reset_backoff(&mut self) {
+        self.backoff = create_backoff(self.client.config.poll_interval_ms)
+    }
+
+    pub async fn sleep_wiggled(&mut self) {
+        let sleepy_time = self.backoff.next().unwrap();
+        debug!(
+            "Client#{} Sleeping for {:?}ms as no results for {}",
+            self.client.id,
+            sleepy_time,
+            self.method.as_str()
+        );
+        tokio::time::sleep(sleepy_time).await;
+    }
+}
+
+pub trait Subscription<ParamType>: Send + Sync + Clone + 'static
+where
+    ParamType: Send + Sync + Serialize + Clone,
+{
+    fn init(&mut self, helper: &SubscriptionHelper) -> Result<(), JsonRpcError>;
+    fn next(&self, helper: &SubscriptionHelper) -> Vec<ParamType>;
+    fn on_send(&mut self, item: &ParamType);
+
+    fn run(mut self, mut helper: SubscriptionHelper) -> Result<JoinHandle<()>, JsonRpcError> {
+        self.init(&helper)?;
+
+        Ok(tokio::spawn(async move {
+            helper.send_ok().await;
+
+            loop {
+                let items = self.next(&helper);
```

Comment:
humor me, can we use a for loop here?

Reply:
I don't believe so; we could `impl Iterator` but given rust breaks out of the `for` when `None` is returned, we'd need another `loop` around the outside of that anyway to ensure we continue to loop forever. Does that make sense?